File: android/core/src/main/kotlin/com/iblocker/core/lists/FilterListUpdater.kt
package com.iblocker.core.lists

import com.iblocker.core.rules.CompiledBlocklist
import com.iblocker.core.rules.SeedRules
import com.iblocker.core.shared.AppPaths
import java.io.IOException
import java.net.ConnectException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import javax.net.ssl.SSLException

data class FetchRequest(
    val url: String,
    val headers: Map<String, String> = emptyMap(),
    val timeoutSeconds: Int = 60,
)

data class FetchResult(
    val statusCode: Int,
    val body: ByteArray,
    val etag: String? = null,
    val lastModified: String? = null,
) {
    override fun equals(other: Any?): Boolean =
        other is FetchResult &&
            statusCode == other.statusCode &&
            body.contentEquals(other.body) &&
            etag == other.etag &&
            lastModified == other.lastModified

    override fun hashCode(): Int {
        var result = statusCode
        result = 31 * result + body.contentHashCode()
        result = 31 * result + (etag?.hashCode() ?: 0)
        result = 31 * result + (lastModified?.hashCode() ?: 0)
        return result
    }
}

/** The network layer, injected so tests never touch the wire. */
fun interface Fetcher {
    suspend fun fetch(request: FetchRequest): FetchResult
}

data class UpdateSummary(
    val updatedSourceIDs: MutableList<String> = ArrayList(),
    val unchangedSourceIDs: MutableList<String> = ArrayList(),
    val failedSourceIDs: MutableMap<String, String> = LinkedHashMap(),
) {
    val anyChanged: Boolean get() = updatedSourceIDs.isNotEmpty()
}

/**
 * Failures we raise ourselves, phrased for the list row a user actually
 * reads rather than for a log.
 */
sealed class FilterListUpdateException(val userMessage: String) : Exception(userMessage) {
    class HttpStatus(val code: Int) : FilterListUpdateException(
        when (code) {
            404 -> "Not found (404) — the list may have moved"
            403 -> "Access denied (403)"
            429 -> "Rate limited — try again later"
            in 500..599 -> "Server error ($code)"
            else -> "Unexpected response ($code)"
        }
    )

    class EmptyBody : FilterListUpdateException("Server returned an empty file")
}

/** Downloads raw list texts into the on-device cache with conditional GETs. */
class FilterListUpdater(
    private val paths: AppPaths,
    private val now: () -> Long = System::currentTimeMillis,
    private val fetcher: Fetcher,
) {

    /**
     * Fetches every enabled source, updating `state.metadata` and the cached
     * list files. Does not recompile; call [BlocklistCompiler] after.
     */
    suspend fun update(state: FilterListState, force: Boolean = false): UpdateSummary {
        val summary = UpdateSummary()
        paths.ensureDirectories()

        for (source in state.sources.filter { it.enabled }) {
            // Bundled rulesets ship in the binary — nothing to fetch.
            if (SeedRules.bundledText(source.id) != null) continue

            var metadata = state.metadata[source.id] ?: FilterListMetadata()
            val headers = LinkedHashMap<String, String>()
            val cacheFile = paths.cachedListFile(source.id)
            if (!force && cacheFile.isFile) {
                metadata.etag?.let { headers["If-None-Match"] = it }
                metadata.lastModified?.let { headers["If-Modified-Since"] = it }
            }

            try {
                val result = fetcher.fetch(FetchRequest(url = source.url, headers = headers))
                when (result.statusCode) {
                    200 -> {
                        if (result.body.isEmpty()) throw FilterListUpdateException.EmptyBody()
                        CompiledBlocklist.writeAtomically(cacheFile, result.body)
                        metadata = metadata.copy(
                            etag = result.etag,
                            lastModified = result.lastModified,
                            lastFetchedMillis = now(),
                            lastError = null,
                        )
                        summary.updatedSourceIDs.add(source.id)
                    }
                    304 -> {
                        metadata = metadata.copy(lastFetchedMillis = now(), lastError = null)
                        summary.unchangedSourceIDs.add(source.id)
                    }
                    else -> throw FilterListUpdateException.HttpStatus(result.statusCode)
                }
            } catch (error: Exception) {
                val message = shortDescription(error)
                metadata = metadata.copy(lastError = message)
                summary.failedSourceIDs[source.id] = message
            }
            state.metadata[source.id] = metadata
        }

        return summary
    }

    private fun shortDescription(error: Throwable): String = when (error) {
        is FilterListUpdateException -> error.userMessage
        is UnknownHostException -> "Can't reach the server"
        is ConnectException -> "Can't reach the server"
        is SocketTimeoutException -> "Timed out"
        is SSLException -> "Secure connection failed"
        is IOException -> error.message?.takeIf { it.isNotBlank() } ?: "Download failed"
        else -> error.message?.takeIf { it.isNotBlank() } ?: "Update failed"
    }
}

File: ios/IBlockerKit/Sources/IBlockerKit/Lists/FilterListUpdater.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public struct FetchResult: Sendable {
    public var statusCode: Int
    public var body: Data
    public var etag: String?
    public var lastModified: String?

    public init(statusCode: Int, body: Data, etag: String? = nil, lastModified: String? = nil) {
        self.statusCode = statusCode
        self.body = body
        self.etag = etag
        self.lastModified = lastModified
    }
}

public struct UpdateSummary: Sendable, Equatable {
    public var updatedSourceIDs: [String] = []
    public var unchangedSourceIDs: [String] = []
    public var failedSourceIDs: [String: String] = [:]

    public var anyChanged: Bool { !updatedSourceIDs.isEmpty }

    public init() {}
}

/// Failures we raise ourselves. Phrased for the list row a user actually
/// reads, not for a log — `URLError`'s own text degrades to
/// "The operation couldn't be completed. (NSURLErrorDomain error -1011.)"
/// when the error is synthesized rather than produced by URLSession.
public enum FilterListUpdateError: LocalizedError, Equatable {
    case httpStatus(Int)
    case emptyBody

    public var errorDescription: String? {
        switch self {
        case .emptyBody:
            return "Server returned an empty file"
        case .httpStatus(let code):
            switch code {
            case 404: return "Not found (404) — the list may have moved"
            case 403: return "Access denied (403)"
            case 429: return "Rate limited — try again later"
            case 500...599: return "Server error (\(code))"
            default: return "Unexpected response (\(code))"
            }
        }
    }
}

/// Downloads raw list texts into the App Group cache with conditional GETs.
/// The network layer is injected as a closure so tests never touch the wire.
public struct FilterListUpdater: Sendable {
    public typealias Fetch = @Sendable (URLRequest) async throws -> FetchResult

    private let fetch: Fetch
    private let paths: AppGroupPaths

    public init(paths: AppGroupPaths, fetch: @escaping Fetch) {
        self.paths = paths
        self.fetch = fetch
    }

    #if canImport(Darwin) || canImport(FoundationNetworking)
    /// Production fetcher backed by URLSession.
    public init(paths: AppGroupPaths, session: URLSession = .shared) {
        self.init(paths: paths) { request in
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                throw URLError(.badServerResponse)
            }
            return FetchResult(
                statusCode: http.statusCode,
                body: data,
                etag: http.value(forHTTPHeaderField: "ETag"),
                lastModified: http.value(forHTTPHeaderField: "Last-Modified")
            )
        }
    }
    #endif

    /// Fetches every enabled source, updating `state.metadata` and the
    /// cached list files. Does not recompile; call BlocklistCompiler after.
    public func update(state: inout FilterListState, force: Bool = false) async -> UpdateSummary {
        var summary = UpdateSummary()
        try? paths.ensureDirectories()

        for source in state.sources where source.enabled {
            // Bundled rulesets ship in the binary — nothing to fetch.
            if SeedRules.bundledText(for: source.id) != nil { continue }
            // A hidden source must not download either: its failures would be
            // recorded where no one can see or act on them.
            if FeatureFlags.isHidden(sourceID: source.id) { continue }

            var metadata = state.metadata[source.id] ?? FilterListMetadata()
            var request = URLRequest(url: source.url)
            request.timeoutInterval = 60
            let hasCachedFile = FileManager.default.fileExists(atPath: paths.cachedListURL(sourceID: source.id).path)
            if !force && hasCachedFile {
                if let etag = metadata.etag {
                    request.setValue(etag, forHTTPHeaderField: "If-None-Match")
                }
                if let lastModified = metadata.lastModified {
                    request.setValue(lastModified, forHTTPHeaderField: "If-Modified-Since")
                }
            }

            do {
                let result = try await fetch(request)
                switch result.statusCode {
                case 200:
                    guard !result.body.isEmpty else { throw FilterListUpdateError.emptyBody }
                    try result.body.write(to: paths.cachedListURL(sourceID: source.id), options: .atomic)
                    metadata.etag = result.etag
                    metadata.lastModified = result.lastModified
                    metadata.lastFetched = Date()
                    metadata.lastError = nil
                    summary.updatedSourceIDs.append(source.id)
                case 304:
                    metadata.lastFetched = Date()
                    metadata.lastError = nil
                    summary.unchangedSourceIDs.append(source.id)
                default:
                    throw FilterListUpdateError.httpStatus(result.statusCode)
                }
            } catch {
                metadata.lastError = shortDescription(of: error)
                summary.failedSourceIDs[source.id] = metadata.lastError
            }
            state.metadata[source.id] = metadata
        }

        return summary
    }

    private func shortDescription(of error: Error) -> String {
        if let updateError = error as? FilterListUpdateError {
            return updateError.errorDescription ?? "Update failed"
        }
        if let urlError = error as? URLError {
            switch urlError.code {
            case .notConnectedToInternet: return "No internet connection"
            case .timedOut: return "Timed out"
            case .cannotFindHost, .cannotConnectToHost, .dnsLookupFailed:
                return "Can't reach the server"
            case .networkConnectionLost: return "Connection lost"
            case .secureConnectionFailed, .serverCertificateUntrusted:
                return "Secure connection failed"
            case .dataNotAllowed: return "Cellular data not allowed"
            case .zeroByteResource: return "Server returned an empty file"
            case .badServerResponse: return "Unexpected response from the server"
            default: return "Download failed (\(urlError.code.rawValue))"
            }
        }
        return error.localizedDescription
    }
}
